chore(OSSF): update token permissions to improve ossf scorecard

laukik-target opened the pull request:
### Description:
This PR aims to improve the security posture of the repository by adhering to the principle of least privilege and enhancing the OSSF Scorecard score. The following changes have been made to reduce excessive permissions in GitHub Actions workflows and ensure a more secure CI/CD pipeline:

Issue: [Bump OSSF score above 9.0 ⬆️](https://github.com/finos/git-proxy/issues/694)

_**Should improve this score:**_
![image](https://github.com/user-attachments/assets/666167a0-2623-4503-b583-b620ec1fca0b)

### Changes:
**1. Updated token permissions:** Scoped permissions at both the top-level and job level for workflows such as:
- ci.yml: Restricted permissions to only contents: read, statuses: write as necessary.
- codeql.yml: Added top-level permissions (contents: read, security-events: write) to ensure CodeQL can upload security events.
- npm.yml: Scoped permissions to packages: write for publishing, with only contents: read for other operations.
- pr-lint.yml: Limited PR-related workflows to pull-requests: write and statuses: write to reduce unnecessary permissions.
- scorecard.yml: Scoped permissions to security-events: write and id-token: write for result publication and scorecard analysis.

**2. Enhanced security:**
- Applied the principle of least privilege to limit the scope of token access.
- Updated all workflows to ensure they use only the permissions necessary for their tasks, preventing unnecessary access to other GitHub components.

**3. Improved OSSF Scorecard rating:**
- These changes directly address issues raised by the OSSF Scorecard tool, enhancing the repository’s security rating by ensuring token permissions follow best security practices.

**Benefits:**
- Reduced risk of unauthorized access due to overly permissive tokens.
- Enhanced overall security posture by minimizing the attack surface in workflows.
- Improved transparency and maintainability by explicitly defining token permissions in each workflow.
- Positive impact on OSSF Scorecard score.

Checklist:
✅Updated workflows with restrictive token permissions.
✅Addressed OSSF Scorecard recommendations for token permissions.

laukik-target commented:
@JamieSlome I am a citi Employee, I Sent an authorization request as a Corporate Contributor for EasyCLA. Can you please approve it?

JamieSlome commented:
@laukik-target - great PR ❤️ Can you re-open using commits created with your Citi e-mail address?

laukik-target commented:
> @laukik-target - great PR ❤️ Can you re-open using commits created with your Citi e-mail address?

I guess, It is good to merge now.

JamieSlome commented on .github/workflows/npm.yml:
```diff
@@ -2,6 +2,12 @@ name: Publish to NPM
+
+permissions:
+  contents: read
+  packages: write
```
```suggestion
```
Not required as we don't use GitHub Packages 👍

laukik-target replied:
removed.

JamieSlome commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "clientinstall": "npm install --prefix client",
     "server": "node index.js",
-    "start": "concurrently 'npm run server' 'npm run client'",
+    "start": "concurrently \"npm run server\" \"npm run client\"",
```
Can we remove from this PR and add to a separate pull request?

laukik-target replied:
Removed from this PR, will be raising new PR for that.

laukik-target commented:
@JamieSlome Changes were done. Requesting for re-review & approval.

laukik-target commented:
@JamieSlome  Please re-review and approve the PR

JamieSlome commented:
@laukik-target - are we able to re-base all of your 10 commits to use your Citi e-mail address instead?

laukik-target commented:
> @laukik-target - are we able to re-base all of your 10 commits to use your Citi e-mail address instead? All 10 commits should use your public Citi e-mail address instead of the current Gmail address 👍 Let me know if you need help on how to do this.

I tried but looks like I messed up something, 

Do you want me to update the email like this?
`git commit --amend --author="Laukik Kakade <laukik.kakade@citi.com>"
`

laukik-target commented:
> @laukik-target - are we able to re-base all of your 10 commits to use your Citi e-mail address instead? All 10 commits should use your public Citi e-mail address instead of the current Gmail address 👍 Let me know if you need help on how to do this.

@JamieSlome  I squashed my commit to single one for clear understanding.
And updated my email to citi email for my single commit but now EasyCLA check has failed.

Please help

laukik-target commented:
> [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/ee533b9dccfeecf8714e32e8f64bd776c29490d5dfa4b9bc60a8b181727cfe6e/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/13399143/256513079/763/#/?version=2) [<img alt="CLA Not Signed" width="328" height="28" src="https://camo.githubusercontent.com/b00e1919a583b4db6c9308e7c0ed62e8bebbb2c573b5a7b520ddd5acf0988f7f/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f742d7369676e65642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/13399143/256513079/763/#/?version=2)
> 
> * ❌ The email address for the commit ([974459d](https://github.com/finos/git-proxy/commit/974459ddf2f224cc24d8c42231cc4d5e489710bd)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://linuxfoundation.atlassian.net/wiki/spaces/LP/pages/160923756/Missing+ID+on+Commit+but+I+have+an+agreement+on+file) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@JamieSlome  What would be the next steps?

laukik-target commented:
@JamieSlome  
PR is ready to merge now.
please check.

laukik-target commented:
@JamieSlome  Can you please merge this PR

laukik-target commented:
@JamieSlome  Can you please approve the PR

laukik-target commented:
@coopernetes  @JamieSlome  

Can you please approve the PR. 
I just fixed a linting issue in last commit.

laukik-target commented:
@JamieSlome  @coopernetes  
Can you please approve this PR

laukik-target commented:
@JamieSlome Can we get this merged

jescalada reviewed (changes requested):
Thanks for the contribution, @laukik-target!

I think your changes on the `scorecard.yml` workflow are making the [workflow error out.](https://github.com/laukik-target/git-proxy/actions/workflows/scorecard.yml)

The OSSF Scorecard forbids write access to `security-events` since it might create backdoors and other security concerns... Write accesses in general should be limited to specific jobs, rather than entire workflows.

Here's [an OSSF blogpost](https://openssf.org/blog/2024/08/12/mitigating-attack-vectors-in-github-workflows/) explaning this in more detail.

This is an example of a successful scorecard workflow run. You can see the `score: 8.1` section which shows the recomputed OSSF score. You can use this to verify whether your changes have improved the score or not:

<img width="2747" height="528" alt="image" src="https://github.com/user-attachments/assets/66b2494e-ad4c-4df9-8a5e-39c0bc40b579" />

Here are some more [successful scorecard runs for finos/git-proxy](https://github.com/finos/git-proxy/actions/workflows/scorecard.yml).

jescalada commented:
@JamieSlome we may want to require the OSSF scorecard CI check before merging... Some kind of conditional check would be even better to prevent accidentally lowering the OSSF score below ~7.5 or so.
